x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IlmHealthIndicatorService.java

```diff
@@ -79,39 +160,174 @@ public HealthIndicatorResult calculate(boolean verbose, int maxAffectedResources
-                    List.of(ImpactArea.DEPLOYMENT_MANAGEMENT)
-                )
-            );
             return createIndicator(
                 YELLOW,
                 "Index Lifecycle Management is not running",
                 createDetails(verbose, ilmMetadata, currentMode),
-                impacts,
+                AUTOMATION_DISABLED_IMPACT,
                 List.of(ILM_NOT_RUNNING)
             );
         } else {
+            return calculateIndicator(verbose, ilmMetadata, currentMode, maxAffectedResourcesCount);
+        }
+    }
+
+    private HealthIndicatorDetails createDetails(boolean verbose, IndexLifecycleMetadata ilmMetadata, OperationMode currentMode) {
+        return createDetails(verbose, ilmMetadata, currentMode, List.of());
+    }
+
+    private HealthIndicatorResult calculateIndicator(
+        boolean verbose,
+        IndexLifecycleMetadata ilmMetadata,
+        OperationMode currentMode,
+        int maxAffectedResourcesCount
+    ) {
+        var stagnatingIndices = stagnatingIndicesFinder.find();
+
+        if (stagnatingIndices.isEmpty()) {
             return createIndicator(
                 GREEN,
                 "Index Lifecycle Management is running",
-                createDetails(verbose, ilmMetadata, currentMode),
+                createDetails(verbose, ilmMetadata, currentMode, stagnatingIndices),
                 Collections.emptyList(),
                 Collections.emptyList()
             );
         }
+
+        return createIndicator(
+            YELLOW,
+            (stagnatingIndices.size() > 1 ? stagnatingIndices.size() + " indices have" : "An index has")
+                + " stayed on the same action longer than expected.",
+            createDetails(verbose, ilmMetadata, currentMode, stagnatingIndices),
+            STAGNATING_INDEX_IMPACT,
+            createDiagnoses(stagnatingIndices, maxAffectedResourcesCount)
+        );
     }
 
-    private static HealthIndicatorDetails createDetails(boolean verbose, IndexLifecycleMetadata metadata, OperationMode mode) {
-        if (verbose) {
-            return new SimpleHealthIndicatorDetails(Map.of("ilm_status", mode, "policies", metadata.getPolicies().size()));
-        } else {
+    private List<Diagnosis> createDiagnoses(List<IndexIlmState> stagnatingIndices, int maxAffectedResourcesCount) {
+        return stagnatingIndices.stream().collect(groupingBy(IndexIlmState::action)).entrySet().stream().map(action -> {
+            var affectedIndices = action.getValue()
+                .stream()
+                .map(IndexIlmState::indexName)
+                .limit(Math.min(maxAffectedResourcesCount, action.getValue().size()))
+                .collect(Collectors.toSet());
+            var affectedPolicies = action.getValue()
+                .stream()
+                .map(IndexIlmState::policyName)
+                .limit(Math.min(maxAffectedResourcesCount, action.getValue().size()))
+                .collect(Collectors.toSet());
```

Comment:
I think we should sort them to ensure the users will not have a stable overview when refreshing the page. What do you think?

Reply:
sure! will update it :)